### PR TITLE
Vastly simplify mailer CDI configuration

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]
Looks good to me. 

We may want to improve the configuration error, as it may just throw an UnresolvedBean exception instead of a proper / understandable error message.

**Comment by geoand** [comment]
Which configuration error are you talking about @cescoffier ?

I also want to do a real test before merging this.

**Comment by cescoffier** [comment]
Right now, if MailSupport cannot be defined or provides broken value, I'm not sure the reporting will work the same way.

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 1]
```diff
@@ -16,52 +20,74 @@
...
+        config.keyStorePassword.ifPresent(cfg::setKeyStorePassword);
+        config.login.ifPresent(s -> cfg.setLogin(LoginOption.valueOf(s.toUpperCase())));
+        config.maxPoolSize.ifPresent(cfg::setMaxPoolSize);
+        config.ownHostName.ifPresent(cfg::setOwnHostname);
+        config.username.ifPresent(cfg::setUsername);
+        config.password.ifPresent(cfg::setPassword);
+        config.port.ifPresent(cfg::setPort);
+        cfg.setSsl(config.ssl);
+        config.startTLS.ifPresent(s -> cfg.setStarttls(StartTLSOptions.valueOf(s.toUpperCase())));
+        cfg.setTrustAll(config.trustAll);
```
Comment (gsmet) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:89]: Let's drop the lambdas while we're at it?

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java [hunk 2]
```diff
@@ -16,52 +20,74 @@
...
+    private MailClient mailClient(Vertx vertx, MailConfig config) {
+        io.vertx.ext.mail.MailConfig cfg = toVertxMailConfig(config);
+        return MailClient.createShared(vertx, cfg);
+    }
+
+    private io.vertx.ext.mail.MailConfig toVertxMailConfig(MailConfig config) {
+        io.vertx.ext.mail.MailConfig cfg = new io.vertx.ext.mail.MailConfig();
+        config.authMethods.ifPresent(cfg::setAuthMethods);
+        cfg.setDisableEsmtp(config.disableEsmtp);
+        cfg.setHostname(config.host);
```
Comment (gsmet) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:77]: So when you add the mailer extension, this is mandatory and things will fail if undefined? It's a bit different than for the other extensions.

Not related to this patch but I wonder if we should revisit this at some point.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added one suggestion and one question. They are not related to the core of the patch though.

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java (hunk 1 above)
Comment (geoand) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:89]: Sure, I'll do that once I get back to this PR :)

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java (hunk 2 above)
Comment (geoand) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:77]: I would prefer not touch that now TBH

File: extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java [hunk 3]
```diff
@@ -75,18 +74,17 @@ void registerAuthClass(BuildProducer<ReflectiveClassBuildItem> reflectiveClass)
...
 
-        RuntimeValue<MailClient> client = recorder.configureTheClient(vertx.getVertx(), beanContainer.getValue(), config,
-                shutdown);
-
-        recorder.configureTheMailer(beanContainer.getValue(), config, launchMode.getLaunchMode());
-
-        return new MailerBuildItem(client);
+        syntheticBeanBuildItemBuildProducer.produce(SyntheticBeanBuildItem.configure(MailerSupport.class)
+                .scope(Singleton.class)
+                .supplier(recorder.mailerSupportSupplier(config, launchMode.getLaunchMode()))
```
Comment (mkouba) [inline extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java:85]: AFAIK it's possible to inject both the config object and the `LaunchMode` so there is no need to register a synthetic bean, or?

Comment (geoand) [reply extensions/mailer/deployment/src/main/java/io/quarkus/mailer/deployment/MailerProcessor.java]: Ah yes, I forgot that `LaunchMode` is also a bean, excellent!

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java (hunk 2 above)
Comment (gsmet) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:79]: Yeah agreed, thus the "at some point". We probably need to discuss that before changing anything.

Comment (geoand) [reply extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailClientProducer.java:79]: :+1:

**Review by gsmet — APPROVED** [review]

**Comment by geoand** [comment]
I'll take this out of draft when I have actually tested it.

**Comment by geoand** [comment]
I tested this and its behavior is the as what we already had.

**Pull request merged**